src/main/ruby/core/regexp.rb

```diff
@@ -157,20 +157,21 @@ def self.union(*patterns)
       enc = convert(patterns.first).encoding
     end
 
-    str = ''.encode(enc)
     sep = '|'.encode(enc)
-    patterns.each_with_index do |pat, idx|
-      str << sep if idx != 0
-      if pat.kind_of? Regexp
-        str << pat.to_s
+    str = ''.encode(enc)
+
+    patterns = patterns.map do |pat|
+      unless pat.kind_of? Regexp
+        Truffle.invoke_primitive(:string_get_rope, StringValue(pat))
```

Comment:
Why are you extracting the Rope here? It's anyway a String or Regexp.

Reply:
To try and make the code at the Java level simpler. If we pass down ropes and regexps then the guard for each cached entry is a simple object comparison (it might miss some ropes that have the same content, but unions are normally pretty static or extremely dynamic). If we pass down `String` then the Java code has to check the type of each array element, and then compare the internals.

Reply:
Seems fairly hidden and complex logic to me.
I would recommend to use a StringEqualNode, since anyway we can't just compare Regexp by identity either.